autoPyTorch/utils/implementations.py

```diff
@@ -61,3 +65,76 @@ def __call__(self, y: Union[np.ndarray, torch.Tensor]) -> np.ndarray:
     @staticmethod
     def get_properties() -> Dict[str, Any]:
         return {'supported_losses': ['BCEWithLogitsLoss']}
+
+
+class MinorityCoalescing(BaseEstimator, TransformerMixin):
+    """ Group together categories which occurence is less than a specified
+    minimum fraction. Coalesced categories get index of one.
+    """
+
+    def __init__(self, minimum_fraction: Optional[float] = None):
+        self.minimum_fraction = minimum_fraction
+
+    def check_X(self, X: np.array) -> None:
+        X_data = X.data if sparse.issparse(X) else X
+        if np.nanmin(X_data) <= -2:
+            raise ValueError("X needs to contain only integers greater than -2.")
+
+    def fit(self, X: np.array, y: Optional[np.ndarray] = None) -> 'MinorityCoalescing':
+        self.check_X(X)
+
+        if self.minimum_fraction is None:
+            return self
+
+        # Remember which values should not be coalesced
+        do_not_coalesce: List[Set[int]] = list()
+        for column in range(X.shape[1]):
+            do_not_coalesce.append(set())
+
+            if sparse.issparse(X):
+                indptr_start = X.indptr[column]
+                indptr_end = X.indptr[column + 1]
+                unique, counts = np.unique(
+                    X.data[indptr_start:indptr_end], return_counts=True)
+                colsize = indptr_end - indptr_start
+            else:
+                unique, counts = np.unique(X[:, column], return_counts=True)
+                colsize = X.shape[0]
+
+            for unique_value, count in zip(unique, counts):
+                fraction = float(count) / colsize
+                if fraction >= self.minimum_fraction:
+                    do_not_coalesce[-1].add(unique_value)
+
+        self.do_not_coalesce_ = do_not_coalesce
+        return self
+
+    def transform(self, X: np.ndarray) -> np.ndarray:
+        self.check_X(X)
+
+        if self.minimum_fraction is None:
+            return X
+
```

Comment:
The codes are a bit complicated and not simple, so I rewrote the codes.
```
def suggestion(X: Union[sparse.csr_matrix, np.ndarray]) -> None:
    n_features = X.shape[1]
    is_sparse = sparse.issparse(X)

    for col in range(n_features):
        if is_sparse:
            indptr_start = X.indptr[col],
            indptr_end = X.indptr[col + 1]
            col_data = X.data[indptr_start:indptr_end]
        else:
            col_data = X[:, col]

        mask = ~np.isin(col_data, do_not_coalesce_)
        col_data[mask] = -2

    return X
```

Just in case, I also checked if it works in the identical way.

```
size = (400000, 100)
n_vals = 10
do_not_coalesce_ = [3, 5]

X = np.random.randint(n_vals, size=size)
X_copy = np.copy(X)

n_wrongs = (original(X) != suggestion(X_copy)).sum()
n_wrongs

>>> 0  # It works identically
```

Plus, I also checked the benchmark speed.

```
size = (400000, 100)
n_vals = 10
do_not_coalesce_ = [3, 5]

X = np.random.randint(n_vals, size=size)
%timeit original(X)
>>> 1.24 s ± 25.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

X = np.random.randint(n_vals, size=size)
%timeit suggestion(X)
>>> 549 ms ± 13.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```